src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -1237,10 +1194,36 @@ void ImageReplayer<I>::start_update_status_task()
 }
 
 template <typename I>
+void ImageReplayer<I>::handle_stop(int r, Context *on_start, Context *on_stop) {
+  {
+    Mutex::Locker locker(m_lock);
+    m_stop_requested = false;
+    m_state = STATE_STOPPED;
+    m_last_r = r;
+    if (r >= 0) {
+      m_state_desc.clear();
+    }
+  }
+  dout(20) << "stop complete" << dendl;
+
+  update_mirror_image_status(true);
+
+  m_local_ioctx.close();
+  m_remote_ioctx.close();
+
+  if (on_start != nullptr) {
+    dout(20) << "on start finish complete, r=" << r << dendl;
+    on_start->complete(r);
+  }
+  if (on_stop != nullptr) {
+    dout(20) << "on stop finish complete, r=" << r << dendl;
+    on_stop->complete(0);
```

Comment:
on_stop is always completed with r=0. It is ok when handle_stop() is called from on_start_fail_finish() (this is previous and expected behaviour, though in the old code the debug message looked wrong, it should have not printed r).

When handle_stop() is called from on_stop_local_image_close_finish() this looks not correct (we lost error code on failed stop).